wdl-lint/src/fix.rs

```diff
@@ -0,0 +1,246 @@
+
+use std::ops::Range;
+use std::ops::RangeInclusive;
+
+use ftree::FenwickTree;
+use serde::Deserialize;
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+/// An insertion point.
+pub enum InsertionPoint {
+    /// Insert immediately before a specified region.
+    BeforeStart,
+    /// Insert immediately after a specified region.
+    AfterEnd,
+}
+
+#[derive(Clone, Debug)]
+/// A replacement to be applied to a String.
+pub struct Replacement {
+    /// The start position of the replacement.
+    start: usize,
+    /// The end position of the replacement.
+    end: usize,
+    /// where to insert the replacement.
+    insertion_point: InsertionPoint,
+    /// Value to be inserted.
+    replacement: String,
+    /// Precedence for replacement. Higher precedences are applied first.
+    precedence: usize,
+}
+
+#[allow(unused)]
+impl Replacement {
+    /// Create a new `Replacement`.
+    pub fn new(
+        start: usize,
+        end: usize,
+        insertion_point: InsertionPoint,
+        replacement: String,
+        precedence: usize,
+    ) -> Self {
+        Replacement {
+            start,
+            end,
+            insertion_point,
+            replacement,
+            precedence,
+        }
+    }
+
+    /// The start position of the replacement.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// The end position of the replacement.
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    /// where to insert the replacement.
+    pub fn insertion_point(&self) -> InsertionPoint {
+        self.insertion_point
+    }
+
+    /// Value to be inserted.
+    pub fn replacement(&self) -> &str {
+        self.replacement.as_ref()
+    }
+
+    /// Precedence for replacement. Higher precedences are applied first.
+    pub fn precedence(&self) -> usize {
+        self.precedence
+    }
+}
+
+// Adapted from ShellCheck's [Fixer](https://github.com/koalaman/shellcheck/blob/master/src/ShellCheck/Fixer.hs)
+/// Apply a series of `Replacement`s to a String.
+///
+/// Internally uses a [Fenwick Tree](https://en.wikipedia.org/wiki/Fenwick_tree)
+/// which is updated as replacements are applied. This allows multiple
+/// replacements referencing only the original input. Although the canonical
+/// Fenwick tree is 1-indexed this implementation is 0-indexed, so replacement
+/// positions must be directly equivalent to string indices.
+#[derive(Clone, Debug)]
+pub struct Fixer {
+    /// The string to be modified.
+    value: String,
+    /// A Fenwick tree for tracking modifications.
+    tree: FenwickTree<i32>,
+}
+
+#[allow(unused)]
+impl Fixer {
+    /// Create a new Fixer from a String.
+    pub fn new(value: String) -> Self {
+        Fixer {
+            tree: FenwickTree::from_iter(vec![0; value.len() + 1]),
+            value,
+        }
+    }
+
+    /// Apply a `Replacement` to the value contained in the Fixer.
+    pub fn apply_replacement(&mut self, rep: &Replacement) {
```

Comment:
```suggestion
    pub fn apply_replacement(&mut self, replacement: &Replacement) {
```
When balancing verbosity VS clarity, our team almost always picks clarity. Long variable names are acceptable round these parts (within reason of course).